Feature/conn

Kevin-Chen0 opened the pull request:
## Proposed changes

Added the following:

- HTTP connection and config: /packages/packages/fetchai/*
- Pet Store YAML file example: /examples/http_ex/*

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

The HTTP connection.py is still largely a WIP with only the baseline functions.

Totoual commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,266 @@
+            protocol=str(envelope.protocol_id),
+            context=b"None",
+            payload=envelope.message,
+        )
```
Keep in mind that if you are using the p2p library from fetch is specific for the p2p implementation that we have in the Ledger constelation. The message in the library is in the form of :
``` 
 msg = {FROM: {NODE_ADDRESS: self.node_address,
                      SENDER_ADDRESS: send_address},
               TO: {NODE_ADDRESS: self.node_address,
                    RECEIVER_ADDRESS: receive_address},
               PROTOCOL: protocol,
               CONTEXT: context,
               PAYLOAD: payload}
```
I mean, I am not sure if this library is useful for what you are doing :)

DavidMinarsch replied:
Agreed, the P2P connection has very little relevance here

DavidMinarsch commented on examples/http_ex/README.md:
```diff
@@ -0,0 +1,9 @@
+# HTTP agent example
```
All documentation should be in the docs folder. Please remove here.

DavidMinarsch commented on packages/fetchai/connections/http/__init__.py:
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2020 Fetch.AI Limited
```
Please make it 2018 - 2020

DavidMinarsch commented on packages/fetchai/connections/http/connection.yaml:
```diff
@@ -0,0 +1,16 @@
+class_name: HTTPConnection
+protocols: [http]
+restricted_to_protocols: ["fetchai/http:0.1.0"]
+excluded_protocols: ["fetchai/http:0.1.0"]
```
this should be empty: `excluded_protocols: []`

DavidMinarsch commented on packages/fetchai/connections/http/connection.yaml:
```diff
@@ -0,0 +1,16 @@
+fingerprint: ""
+description: "The HTTP connection that wraps a RESTful API specification."
+class_name: HTTPConnection
+protocols: [http]
```
this should be `protocols: ["fetchai/http:0.1.0"]`

DavidMinarsch commented on packages/fetchai/connections/http/connection.yaml:
```diff
@@ -0,0 +1,16 @@
+  port: ${port:8000}
+  api: '' # put here the dotted path to your API spec name.
+dependencies:
+  api: {}
```
please remove `api: {}`, it should be `dependencies: {}`

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,266 @@
+
+    def __init__(self,
+                 address: Address,
+                 api_spec: str,  # In JSON serialized form, might switch to yaml or server_stub later
```
The swagger spec above is in yaml, so why not go with a path to that file: `api_spec_path:`

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,266 @@
+    def __init__(self,
+                 address: Address,
+                 api_spec: str,  # In JSON serialized form, might switch to yaml or server_stub later
+                 provider_addr: str,
```
what are `provider_addr` and `provider_port` used for?

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,266 @@
+    """Proxy to the functionality of the RESTful API."""
+
+    def __init__(self,
+                 address: Address,
```
what is address used for?

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,266 @@
+        """
+        Initialize a connection to an RESTful API.
+
+        :param address: the address used in the protocols.
```
Not matching the function signature

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,266 @@
+        """
+
+        # the following api_spec format checks will be in their own function check_api(api_spec)
+        try:
```
We should load a yaml file (see here for an example: https://github.com/fetchai/agents-aea/blob/master/aea/configurations/loader.py#L81)

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,266 @@
+            :api_spec - configuation of setting-up a RESTful-to-envelope messaging with given skill or other core component
+        :return: the connection object
+        """
+        addr = cast(str, connection_configuration.config.get("addr"))
```
see above

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,266 @@
+
+        :param address: the address
+        """
+        self.address = address
```
note, in the P2PChannel, the P2PChannel acts as the client. Here, the HTTPChannel acts as a server.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,266 @@
+        """
+        with self.lock:
+            if self.stopped:
+                self._httpCall = HTTPCalls(
```
not relevant, as this is supposed to be the server

DavidMinarsch replied:
to start with, you could make HTTPChannel inherit from https://docs.python.org/2/library/basehttpserver.html#BaseHTTPServer.BaseHTTPRequestHandler
and here an example https://wiki.python.org/moin/BaseHttpServer. Then you just need to implement the `do_..` methods. i.e. do_GET, do_HEAD, do POST and create the envelopes which you hand off to the Connection

DavidMinarsch replied:
this isn't ready for production but it is ok for the demo

Kevin-Chen0 replied:
Thanks for that, I will check out BaseHTTPServer.

DavidMinarsch commented:
1. Yes please move `examples/http_ex/README.md`. In examples we just keep readmes for things not worth mentioning in the docs. So `gym` example can stay there.
2. agreed, would change to host and port.
3. we should only add functionality we know we need. Atm we don't seem to need it so let's remove it.

Kevin-Chen0 commented:
**Updates:**
1) Fixed the /http/connection.yaml file.
2) Updated /http_ex/apis/petstore.yaml file to latest OpenAPI v3.0.2 one found here: http://petstore.swagger.io:8080/.
3) Added the /http_ex/apis/petstore_sim.yaml file, which is simplified version of petstore.yaml used for HTTPConntection testing. 
4) Wrote two functions in /http/connection.py that convert the paths in api_spec into Envelopes each. Although the API responses in JSON format is the Envelope message atm, that may change on how the Envelope receives handles it. The full API path is specified in the uri of the Envelope.
5) Added HTTP Server and Handler in connection.py that replaces httpCalls from P2P. 

**Next steps:**
1) Continue populating the HTTPHandler.
2) Begin drafting the _petstoresim_ Skill using the petstore_sim.yaml.
3) Write the send/receiving Envelop functions in conntection.py between AEA and skill. This is for when the front end client is making requests.
4) Test using CLI tool, such as between two different computers. One for front-end client and other for HTTP server AEA.
5) Update the docs.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+    """Proxy to the functionality of the web RESTful API."""
+
+    def __init__(self,
+                 # address: Address,
```
Please remove all unnecessary codes, comments should be used sparingly.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+
+    def __init__(self,
+                 # address: Address,
+                 api_path: str,  # Directory path of the API YAML file.
```
Please provide this description in the docstring of the parameter or make the parameter name more explicit.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+        if kwargs.get("connection_id") is None:
+            kwargs["connection_id"] = PublicId("fetchai", "http", "0.1.0")
+        super().__init__(*args, **kwargs)
+        # self.address = address
```
see above

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+        """
+        addr = cast(str, connection_configuration.config.get("host"))
+        port = cast(int, connection_configuration.config.get("port"))
+        api_path = cast(str, connection_configuration.config.get("api"))
```
Please use same var name in config if possible

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+            p.name for p in connection_configuration.excluded_protocols
+        }
+        return HTTPConnection(
+            # address,
```
see above

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+        path_split = path_name.split(separator)
+
+        to = path_split[1].strip()  # to: name of Skill (petstoresim)?
+        sender = 'TBD'  # hardcoded for now
```
One way would be to create a uuid for each request. Then the sender is the uuid. You send the envelope into the agent, whilst blocking on the response to the request. Once a matching envelope with the same uuid as received comes back from within the agent (or a timeout is hit) your serve the response.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+        protocol_id = PublicId.from_str('fetchai/http:0.1.0')  # hardcoded for now
+        uri = URI(f"http://{self.host}:{self.port}{path_name}")
+        context = EnvelopeContext(connection_id=self._connection_id, uri=uri)
+        # Msg consists of all of the responses for each REST req in JSON format.
```
not sure about this

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+    def _process_api(self, api_spec):
+        """Process the api_spec.
+
+        Decode all paths to get envelops each, and put it in the agent's inbox.
```
not sure why you;d want to do that.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+
+    def do_GET(s):
+        """Respond to a GET request."""
+        # Placeholder functions
```
hear is where you'd create the envelope, send it in to the agent, wait for the response envelope and then serve the response

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+
+    def do_POST(s):
+        """Respond to a GET request."""
+        pass
```
here is where you'd create the envelope, sent it to the agent and serve a response (potentially without waiting for response from agent)

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,359 @@
+        except OSError:
+            logger.error(f"{host}:{port} is already in use, please try another Socket.")
+        # Process API specifications into Envelopes to be sent to InBox
+        self._process_api(api_spec)
```
I would use the api_spec to filter for valid paths which the handlers should process

DavidMinarsch commented:
> **Updates:**
> 
> 1. Fixed the /http/connection.yaml file.
> 2. Updated /http_ex/apis/petstore.yaml file to latest OpenAPI v3.0.2 one found here: http://petstore.swagger.io:8080/.
> 3. Added the /http_ex/apis/petstore_sim.yaml file, which is simplified version of petstore.yaml used for HTTPConntection testing.
> 4. Wrote two functions in /http/connection.py that convert the paths in api_spec into Envelopes each. Although the API responses in JSON format is the Envelope message atm, that may change on how the Envelope receives handles it. The full API path is specified in the uri of the Envelope.
> 5. Added HTTP Server and Handler in connection.py that replaces httpCalls from P2P.
> 6. Other minor changes, including renaming to _host_ and _port_.
> 
> **Next steps:**
> 
> 1. Continue populating the HTTPHandler.
> 2. Begin drafting the _petstoresim_ Skill using the petstore_sim.yaml.
> 3. Write the send/receiving Envelop functions in conntection.py between AEA and skill. This is for when the front end client is making requests.
> 4. Add /agents/http_server in the packages with HTTP Connection, Protocol, and example [petstoresim] Skill bundled together and can be easily fetched.
> 5. Test http_server AEA using CLI tool, such as between two different computers. One for front-end client and other for HTTP server AEA.
> 6. Update the docs.

thoughts:

2. seems a bit excessive, keep it simple stupid :)
3. yes, simple!
4. not quite sure what you're trying to do there and why; see my comments

rennext steps:
1. good idea, see my comments for suggestions
2. I'd put this on hold until the connection is finished.
3. this should happen in the handler mostly imo
4. not a priority imo
5. for now I'd focus on the connection and writing unit tests for it
6. especially for the connection, this would be helpful

Kevin-Chen0 commented:
**Updates:**

1) Completed the do_GET and do_POST functions in HTTPHandler so that they can be pushed to the _process_request() function in HTTP channel.
2) Completed the _process_request() function that packages the request (from HTTPHandler) into Envelope based on DavidM's specs and sent _in_queue_ to the Agent.
3) Successfully tested HTTP Connection (to Envelope part) using a GET and a POST query from pet store, as specified in OpenAPI Specs I/O.

**Next steps:**

1) Complete the _send_response() that sends the response back to the front-end client, given the _out_queue_ Envelope from the Agent. I think that this is the last core part of HTTP Connection.
2) Write a preliminary from of Echo Skill handler that receives the _in_queue_ Envelopes and write the response Envelope _out_queue_, which will finally be handled by _send_response().
3) Updates the docs when the following two steps are complete.

DavidMinarsch commented:
Cancelling to be able to see the changes on a new PR against develop
